### DISCUSSION TITLE
Automating clang-format and clang-tidy targets

### DISCUSSION THREAD

**Discussion by junglie85** [thought-machine/please#2785] [General]
I've liked the look of please for a while but have been working on Windows so it's not been possible to try it out. Now I'm back in the *nix world, I'm looking forward to getting my hands dirty with it.

I'm working with a very simple hello world in c++ and would like to do the following:

- Use a system installed version of `clang-format` and configure please to manually invoke it via a target.
- Use a system installed version of `clang-tidy` and configure please to manually invoke it via a target.
- Configure please to automatically invoke these tools as part of the build process for a `cc_binary` target.

For `clang-format`, I need to [find all relevant files](https://stackoverflow.com/questions/28896909/how-to-call-clang-format-over-a-cpp-project-folder), which gives me a task a bit like this:

```
sh_cmd(
    name = "clangformat",
    cmd = "find . -iname *.h -o -iname *.cpp | xargs clang-format -i",
)
```

I'd rather something like:

```
sh_cmd(
    name = "clangformat",
    cmd = "clang-format -i **/*.cpp" **/*.h,
)
```

But that fails if there are no header files. How can I use please to recursively find all `c,cpp,cxx,h,hpp,hxx,inl` etc files and pass them as arguments to `clang-format`?

Similarly, how can I do this for `clang-tidy`?

```
sh_cmd(
    name = "clangtidy",
    cmd = "clang-tidy **/*.cpp -- **/*.cpp",
)
```

`clang-tidy` works much better with a compilation database. CMake can generate a compilation database; is this something please is able to do or in the roadmap if not currently possible?

How do I link these `sh_cmd`s to the `cc_binary` target as they're not tools in the `tools` folder?

**Comment by Tatskaari** [comment]
The short answer is no, we don't have that tight integration with the C/C++ ecosystem. As you've discovered, you can use Please as a task runner to run commands via `sh_cmd()`. You can also use commands like `plz query inputs`, `plz query print --field srcs`, or `plz query graph` to get information about the build graph including the inputs to each target, but they won't directly give you what you're after. You may be able to build a compilation database off the back of these tools though. 

Another approach that might work would involve extending the C/C++ rules. Similar to how the Golang Please package driver works, we could output the compilation database info for each cc_library() rule as a sub-rule. For context, Golang has this concept of a package json file that describes something very similar i.e. all the source files in a particular translation unit which can be used by other tools to understand foreign build systems like Please. We have implemented this in the go rules by adding an internal `#pkg_info` target to the `go_library()` rules that outputs a json file describing things about the package, including the source files. A similar approach for C/C++ rules might make sense. 

By the way, there was a bit of discussion a while back about linters here, which you may find interesting:
https://github.com/thought-machine/please/issues/1294

Nothing really happened off the back of this, but the idea was that rules would define lint actions so you'd just be able to `plz lint //src/...` and it would run `clang-tidy` because `cc_library()` etc. define that as a lint action. Unfortunately a lot of linters require access to the whole source graph, and we never really came up with an elegant design that would work for this use case.
